packages/backend/src/core/VerifierCollector.ts

```diff
@@ -0,0 +1,112 @@
+import { utils } from 'ethers'
+import { AbiCoder } from 'ethers/lib/utils'
+
+import { EthereumAddress } from '../model'
+import {
+  ImplementationAddedEventRecord,
+  UpgradedEventRecord,
+  VerifierEventRecord,
+  VerifierEventRepository,
+} from '../peripherals/database/VerifierEventRepository'
+import { EthereumClient } from '../peripherals/ethereum/EthereumClient'
+import { BlockRange } from '../peripherals/ethereum/types'
+import { partition } from '../tools/partition'
+
+const PROXY_ADDRESS = '0xC8c212f11f6ACca77A7afeB7282dEBa5530eb46C'
+
+const PROXY_ABI = new utils.Interface([
+  'event ImplementationAdded(address indexed implementation, bytes initializer, bool finalize)',
+  'event Upgraded(address indexed implementation)',
+])
+
+const Upgraded = PROXY_ABI.getEventTopic('Upgraded')
+const ImplementationAdded = PROXY_ABI.getEventTopic('ImplementationAdded')
+
+export class VerifierCollector {
+  constructor(
+    private readonly ethereumClient: Pick<EthereumClient, 'getLogs'>,
+    private readonly verifierEventRepository: VerifierEventRepository
+  ) {}
+
+  /**
+   * Saves new `Upgraded` and `ImplementationAdded` events to the database,
+   * and returns verifier addresses derived from both the old and new events.
+   */
+  async collect(blockRange: BlockRange): Promise<EthereumAddress[]> {
+    const oldEvents = this.verifierEventRepository.getAll()
+    const newEvents = await this.getEvents(blockRange)
+    const savingNewEventsToDb =
+      this.verifierEventRepository.addOrUpdate(newEvents)
+
+    const events = [...(await oldEvents), ...newEvents]
+    const [upgraded, added] = partitionVerifierEvents(events)
+
+    added.sort(byBlockNumberDescending)
+
+    await savingNewEventsToDb
+
+    return computeVerifiersFromEvents(added, upgraded)
```

Comment:
This could be simplified. If `getEvents` also called awaited `verifierEventRepository.addOrUpdate` you could write:
```ts
const [oldEvents, newEvents] = await Promise.all([
  this.verifierEventRepository.getAll(),
  this.getEvents(blockRange)
])
```
And the code would be equivalent with the small exception that partitioning and computing happen after the events are persisted, which shouldn't make a noticeable difference

Reply:
I'm a bit unsure about persisting events to database happening in method called `getEvents`, thus "collect".

If you feel that this late `await savingNewEventsToDb` is harming readability, I can await it where `add` is called.